dlib/matrix/matrix_fft.h

```diff
@@ -369,31 +386,142 @@ namespace dlib
 
             const size_t ntime = (stft.nr() - 1) * hoplen + wlen;
             matrix<ReturnType> signal = zeros_matrix<ReturnType>(1, ntime);
-            matrix<R> norm   = zeros_matrix<R>(1, ntime);
-            matrix<R> win    = window(ones_matrix<R>(1, wlen), w, PERIODIC, {});
-            matrix<R> win2   = squared(win);
+            matrix<R> norm = zeros_matrix<R>(1, ntime);
+            matrix<R> win(1, wlen);
+            matrix<R> win2(1, wlen);
+            for (std::size_t i = 0 ; i < wlen ; ++i) {
+                win(0, i)   = w(i, wlen, type_t<R>{});
+                win2(0, i)  = win(0, i) * win(0, i);
+            }
 
             for (long t = 0 ; t < stft.nr() ; ++t)
             {
                 set_subm(signal, 0, t*hoplen, 1, wlen) += pointwise_multiply(win, subm(ifft_obj(rowm(stft, t)), 0, 0, 1, wlen));
                 set_subm(norm,   0, t*hoplen, 1, wlen) += win2;
             }
 
-            /*! Remove padding of wlen/2 and wlen on either end !*/
-            norm = subm(norm, 0, wlen/2, 1, ntime - wlen);
-            DLIB_ASSERT(sum(norm < 1e-13) == 0, "NOLA constraint not satisfied");
-            signal = pointwise_divide(subm(signal, 0, wlen/2, 1, ntime - wlen), norm);
+            /*! Remove padding of wlen/2 and wlen/2 on either end !*/
+            DLIB_ASSERT(sum(subm(norm, 0, wlen/2, 1, ntime - wlen) < 1e-13) == 0, "NOLA constraint not satisfied");
+            signal = pointwise_divide(subm(signal, 0, wlen/2, 1, ntime - wlen),
+                                      subm(norm,   0, wlen/2, 1, ntime - wlen));
 
             return signal;
         }
     }
 
 // ----------------------------------------------------------------------------------------
 
-    template <typename EXP>
+    struct hann_window
+    {
+        /*!
+            WHAT THIS OBJECT REPRESENTS
+                This returns a function object with signature R(size_t i, size_t N) which computes
+                a periodic Hann window which can be passed to the STFT family of functions.
+        !*/
+
+        template<typename R>
+        R operator()(std::size_t i, std::size_t N, type_t<R>) const
+        /*!
+            requires
+                - R is float, double, or long double
+        !*/
+        { return hann<R>(i, N, PERIODIC);};
```

Comment:
```suggestion
        double operator()(std::size_t i, std::size_t N) const
        { return hann(i, N, PERIODIC);};
```

How about this?   Like in general don't template on return type and just make everything return `double`.  Consider the hann code:
```
        const R phi = (2.0 * pi * i) / size;
        return 0.5 - 0.5 * std::cos(phi);
```
`(2.0 * pi * i) / size` is a `double`, which if `R` is `float` gets downcast to `float`, then the `float` version of `cos` runs on it and gives a less accurate result than it could have.  And then it gets converted back to `double` again by `0.5 *` before being cast back to `float` implicitly.   And `float` isn't faster than `double` on most CPUs.  So if this is all just `double` internally people will get more accurate results.  They can still stick it in a `matrix<float>` if they want certainly, but letting the cast to `float` only happen one at the very end will make the resulting value more accurate.  As well as simplifying the code here too.  Might even run faster since a bunch of float/double conversions would be removed.